src/classes/Collection.cls

```diff
@@ -2,317 +2,494 @@ public with sharing class Collection {
-		for (SObject rec : records) {
-			Id key = (Id) rec.get(field);
-			if (!grouped.containsKey(key)) {
-				grouped.put(key, (List<SObject>) listType.newInstance());
-			}
-			grouped.get(key).add(rec);
-		}
-		return grouped;
+
+	public Map<Decimal, List<SObject>> groupByDecimals(SObjectField field, Type listType) {
+		return groupByDecimals(''+field, listType);
+	}
+
+
+	public Map<Decimal, List<SObject>> groupByDecimals(String field, Type listType) {
+		Map<Object, List<SObject>> result = groupByObjects(field, Decimal.class, listType);
+		return (Map<Decimal, List<SObject>>) JSON.deserialize(JSON.serialize(result), Map<Decimal, List<SObject>>.class);
+	}
+
+
+	public Map<Id, List<SObject>> groupByIds(SObjectField field) {
+		return groupByIds(''+field);
 	}
 
-	public Map<Id, List<SObject>> groupByIds(Schema.SObjectField field) {
+
+	public Map<Id, List<SObject>> groupByIds(String field) {
 		return groupByIds(field, List<SObject>.class);
 	}
 
-	public Map<String, List<SObject>> groupByStrings(Schema.SObjectField field, Type listType) {
-		Map<String, List<SObject>> grouped = new Map<String, List<SObject>>();
-		for (SObject rec : records) {
-			String key = (String) rec.get(field);
-			if (!grouped.containsKey(key)) {
-				grouped.put(key, (List<SObject>) listType.newInstance());
-			}
-			grouped.get(key).add(rec);
-		}
-		return grouped;
+
+	public Map<Id, List<SObject>> groupByIds(SObjectField field, Type listType) {
+		return groupByIds(''+field, listType);
+	}
+
+
+	public Map<Id, List<SObject>> groupByIds(String field, Type listType) {
+		Map<Object, List<SObject>> result = groupByObjects(field, Id.class, listType);
+		return (Map<Id, List<SObject>>) JSON.deserialize(JSON.serialize(result), Map<Id, List<SObject>>.class);
+	}
+
+
+	public Map<String, List<SObject>> groupByStrings(SObjectField field) {
+		return groupByStrings(''+field);
 	}
 
-	public Map<String, List<SObject>> groupByStrings(Schema.SObjectField field) {
+
+	public Map<String, List<SObject>> groupByStrings(String field) {
 		return groupByStrings(field, List<SObject>.class);
 	}
 
-	public Collection pick(List<Schema.SObjectField> fields) {
-		return pick(new Set<Schema.SObjectField>(fields));
+
+	public Map<String, List<SObject>> groupByStrings(SObjectField field, Type listType) {
+		return groupByStrings(''+field, listType);
+	}
+
+
+	public Map<String, List<SObject>> groupByStrings(String field, Type listType) {
+		Map<Object, List<SObject>> result = groupByObjects(field, String.class, listType);
+		return (Map<String, List<SObject>>) JSON.deserialize(JSON.serialize(result), Map<String, List<SObject>>.class);
 	}
 
-	public Collection pick(Set<Schema.SObjectField> fields) {
-		Set<String> fieldNames = new Set<String>();
-		for (Schema.SObjectField field : fields) {
-			Schema.DescribeFieldResult describe = field.getDescribe();
-			fieldNames.add(describe.getName());
+
+	public Map<Object, List<SObject>> groupByObjects(String field, Type objectType, Type listType) {
```

Comment:
This is very elegant in terms of code used. I believe that it's very handy for field types that are not covered by the library. However, being that this is a _library_, I'm concerned that it is an pointless performance hit to use it for primitives in the name of elegance. The benefit of doing that is the removal of 100 lines of code from a library.

The drawback for a list with N records, is N single record serializations, N single record deserializations, and finally a JSON serialization and deserialization of the entire list. We're looking at ~4N record serialization operations which are not required.

My suggestion is to:

- revert all the changes to `groupBy*`
- consider introducing `groupByType` (as opposed to `groupByObjects`) which will cover types not otherwise supported, at the expense of loss of some type safety (for Map keys)